### PR TITLE
Feature/dato panel refactor

### PR DESCRIPTION
**Pull request by shaharkazaz**
Docs will be updated by me today/tomorrow and pushed separately.

BREAKING CHANGES: The dato panel API changed

### CODE REVIEW

File: lib/src/panel/panel-ref.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  /**
+   * Gets an observable that is notified when the panel is finished closing.
+   */
+  afterClosed(): Observable<any> {
+    return this._onClose.asObservable().pipe(take(1));
+  }
+
+  __destroy() {
```
Comment (NetanelBasal) [inline lib/src/panel/panel-ref.ts]: one underscore is sufficient for privates.

File: lib/src/panel/panel.service.ts [hunk 2]
```diff
@@ -5,111 +5,157 @@
...
+  close(result?: any): void {
+    this.tryClose({
+      trigger: PanelResultCloseTrigger.CLOSE,
+      data: result
+    }).subscribe();
+  }
+
+  private tryClose(result: DatoPanelResult): Observable<void> {
+    if (this.panelsStack.length) {
+      const lastPanel = this.panelsStack[this.panelsStack.length - 1];
```
Comment (NetanelBasal) [inline lib/src/panel/panel.service.ts]: use `last` from `datorama/utils`

File: lib/src/panel/panel.service.ts [hunk 3]
```diff
@@ -5,111 +5,157 @@
...
 
+    /* Events Registration*/
+    this.registerToEvents(config, relativeTo);
+  }
+
+  /**
+   * Closing the panel with an optional result object
+   * @param result
+   */
+  close(result?: any): void {
```
Comment (NetanelBasal) [inline lib/src/panel/panel.service.ts]: generic, not any.

File: lib/src/panel/panel-ref.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+   * Check if the panel can be closed
+   * @param result
+   * @return {Observable<boolean>}
+   */
+  private canClose(result: DatoPanelResult): Observable<boolean> {
+    if (!this._beforeClose.length) {
+      return of(true);
+    }
+
+    const observers$ = [];
```
Comment (NetanelBasal) [inline lib/src/panel/panel-ref.ts]:
```ts 
const observers$  =  this._beforeClose.map(handler => {
   let value = handler(result);
   return isBoolean(value) ? of(value) : value;
})

 return forkJoin(observers$).pipe(
    map(resultArr => resultArr.every(res => toBoolean(res)))
);
``

File: lib/src/panel/panel-ref.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    this._onClose.next(this._result);
+    this._onClose.complete();
+  }
+
+  /**
+   * Check and close the panel
+   * @param {DatoPanelResult} result
+   */
+  __tryClose(result: DatoPanelResult): Observable<void> {
+    return new Observable(observer => {
```
Comment (NetanelBasal) [inline lib/src/panel/panel-ref.ts]:
```ts
return this.canClose(result).pipe(
  switchMap(canClose => {
   if(!canClose) {
      return throwError()
   }
  this._result = result;
  /* Signal to the service to start the slideOut animation and destroy completely the panel */
  this._onDestroy.next();
  this._onDestroy.complete();
  return this.afterClosed();
})
)
```

File: lib/src/panel/panel.service.ts [hunk 6]
```diff
@@ -5,111 +5,157 @@
...
-      this.close();
-      this.destroy$.pipe(take(1)).subscribe(() => {
-        this.createPanel(content, _relativeTo, options);
-      });
+    if (this.panelContainer && this.isOpen) {
+      this.tryClose({
+        trigger: PanelResultCloseTrigger.SWITCH_CONTENT
+      }).subscribe(
+        () => {
+          /* The previous panel was closed successfully.
```
Comment (NetanelBasal) [inline lib/src/panel/panel.service.ts]: fix alignment

File: lib/src/panel/panel-ref.ts (hunk 4 above)
Comment (shaharkazaz) [reply lib/src/panel/panel-ref.ts]: @NetanelBasal where is the isBoolean imported from?

**Pull request merged**